[Assist] Add AssistAction and AssistAccessRequest PostHog event reporting

jakule opened the pull request:
This commit introduced new usage tracking for Assist features. AssistAction event is emitted when user triggers an action like SSH command generation or output explaining. AssistAccessRequest event is fired when user requests access to a resource. The events capture anonymized username, resource type or action triggered, and a breakdown of tokens used for generating command summaries or completion responses.

jakule committed:
Add AssistAction and AssistAccessRequest PostHog event reporting

This commit introduced new usage tracking for Assist features. AssistAction event is emitted when user triggers an action like SSH command generation or output explaining. AssistAccessRequest event is fired when user requests access to a resource. The events capture anonymized username, resource type or action triggered, and a breakdown of tokens used for generating command summaries or completion responses.

hugoShaka commented on api/proto/teleport/usageevents/v1/usageevents.proto:
```diff
@@ -377,6 +377,36 @@ message AssistNewConversationEvent {
+// AssistAccessRequest is an event that is emitted when a user requests access
+// to a resource via Assist.
+message AssistAccessRequest {
+  // UserName is anonymized user name
+  string user_name = 1;
```
There is no need to put the `user_name` field in this message, we should remove it as the information is already carried by the certificate used to connect to the auth.


<details>
  <summary>boring details</summary>
The messages in `api/proto/teleport/usageevents/v1/usageevents.proto` are here for the various clients (tbot, discover, proxy, ...) to report events to the auth. Then, the auth emits batch of events to prehog using messages from `lib/proto/prehog/v1alpha1/teleport.proto`. Those events contain more context, including cluster_name, license_name, and user_name (user_name is event-sepcific while cluster_name and license_name are on the generic event part).

This metadata is filled by the auth itself (user_name gets populated from the user metadata extracted from the certificate used to speak to the auth, this happens [here](https://github.com/gravitational/teleport/pull/30513/files#diff-f9d80bd23db506a0a06486e2812e59ebd25531ee5246d3476278dde911922a87R383-R384)). This avoids data duplication and ensure no mistake can be made when reporting an event (and no client can lie about its metadata).
</details>

jakule replied:
Removed

hugoShaka commented on lib/web/assistant.go:
```diff
@@ -520,13 +525,29 @@ func (h *Handler) assistSSHExplainOutputLoop(ctx context.Context, assistClient *
+			},
+		},
+	}
+	if err := usageRep.SubmitUsageEvent(ctx, usageEventReq); err != nil {
+		h.log.WithError(err).Warn("Failed to emit usage event")
+	}
```
Should we report events synchronously? If somehting bad happens and this hangs or is slow, this will block execution. Also, if the context gets cancelled (e.g. the client disconnected) we won't report the event.

This discussion is ongoing here: https://github.com/gravitational/teleport/pull/30554#issuecomment-1682319427

Until a consensus is reached, I think we should align all assist reporting calls. Currently most of them are asynchornous with their own background context, but those new calls are synchronous.

jakule replied:
I made them synchronous for a different reason. I don't think we should wait for the request to be sent, as you mentioned, but this call requires auth client. Auth client can be closed as soon as the connection closes, so we may lose many events. 
I agree that we should agree on something, but I also don't think this is a blocker for the current implementation.

avatus commented on lib/usagereporter/teleport/types.go:
```diff
@@ -618,6 +618,38 @@ func (e *AssistNewConversationEvent) Anonymize(a utils.Anonymizer) prehogv1a.Sub
 
+type AssistAccessRequestEvent prehogv1a.AssistAccessRequestEvent
+
+func (e *AssistAccessRequestEvent) Anonymize(a utils.Anonymizer) prehogv1a.SubmitEventRequest {
```
Does this need some docs?

jakule replied:
Added

avatus reviewed (approved):
+1 on not needing the username but rest looks fine to me

jakule committed:
"Removed userName in AssistAccessRequest and AssistAction"

Both "AssistAccessRequest" and "AssistAction" messages in usageevents.proto file and their corresponding Go structures were modified to remove the "userName" field.

jakule committed:
Add GoDoc
